subprojects/maven/src/main/java/org/gradle/api/publish/maven/internal/publication/DefaultMavenPublication.java

```diff
@@ -108,26 +123,34 @@ public void from(SoftwareComponent component) {
         this.component = (SoftwareComponentInternal) component;
 
         Set<PublishArtifact> seenArtifacts = Sets.newHashSet();
-        for (UsageContext usageContext : this.component.getUsages()) {
+        Set<ModuleDependency> seenDependencies = Sets.newHashSet();
+        for (UsageContext usageContext : getUsageContexts()) {
             // TODO Need a smarter way to map usage to artifact classifier
             for (PublishArtifact publishArtifact : usageContext.getArtifacts()) {
-                if (!seenArtifacts.contains(publishArtifact)) {
-                    seenArtifacts.add(publishArtifact);
+                if (seenArtifacts.add(publishArtifact)) {
                     artifact(publishArtifact);
                 }
             }
 
             Set<MavenDependencyInternal> dependencies = dependenciesFor(usageContext.getUsage());
             for (ModuleDependency dependency : usageContext.getDependencies()) {
-                if (dependency instanceof ProjectDependency) {
-                    addProjectDependency((ProjectDependency) dependency, dependencies);
-                } else {
-                    addModuleDependency(dependency, dependencies);
+                if (seenDependencies.add(dependency)) {
+                    if (dependency instanceof ProjectDependency) {
+                        addProjectDependency((ProjectDependency) dependency, dependencies);
+                    } else {
+                        addModuleDependency(dependency, dependencies);
+                    }
                 }
             }
         }
     }
 
+    private List<UsageContext> getUsageContexts() {
```

Comment:
It's probably a matter of taste but I don't like getters which are not getters. I would prefer a name like `sortedUsageContexts()`